File: src/main/java/VacationType.java

import java.util.ArrayList;
import java.util.List;

public class VacationType
{
	public int rangeMin;
    public int rangeMax;
    public int startMin;
    public int stopMax;
    public int breakDurationMin;
    public int breakDurationMax;
    public int startToBreakMin;
    public int breakToStopMin;
    public int limit;
    public int cost;
    
    public Vacation[] enumerate()
    {
    	List< Vacation > list = new ArrayList< Vacation >();
    	Vacation vacation = new Vacation();
        vacation.cost = cost;
        for(int range = rangeMin; range <= rangeMax; ++range)
        for(int start = startMin; start + range <= stopMax + 1; ++start)
        {
            vacation.start = start;
            vacation.stop = start + range - 1;
            for(int breakDuration = breakDurationMin; breakDuration <= breakDurationMax; ++breakDuration)
            for(int breakStart = vacation.start + startToBreakMin; breakStart + breakDuration + breakToStopMin <= vacation.stop + 1; ++breakStart)
            {
                vacation.breakStart = breakStart;
                vacation.breakStop = breakStart + breakDuration - 1;
                list.add(new Vacation(vacation));
            }
        }
        Vacation[] result = new Vacation[list.size()];
        int typeIndex = 0;
        for(Vacation element : list)
        {
        	result[typeIndex++] = element;
        }
        return result;
    }
    
    public boolean check()
    {
    	return
    		   (rangeMin > 0)
    		&& (rangeMax > 0)
    		&& (startMin > 0)
    		&& (stopMax > 0)
    		&& (breakDurationMin > 0)
    		&& (breakDurationMax > 0)
    		&& (startToBreakMin > 0)
    		&& (breakToStopMin > 0)
    		&& (rangeMin <= rangeMax)
    		&& (startMin + rangeMax <= stopMax + 1)
    		&& (breakDurationMin <= breakDurationMax)
    		&& (startToBreakMin + breakDurationMin + breakToStopMin <= rangeMin)
    		&& (startToBreakMin + breakDurationMax + breakToStopMin <= rangeMax);
    }
}

File: cplusplus/vacation_type.cpp
#include "vacation_type.h"

using namespace std;

VacationType::VacationType(const unsigned int durationMin, const unsigned int durationMax, const unsigned int startMin, const unsigned int stopMax, const unsigned int breakDurationMin, const unsigned int breakDurationMax, const unsigned int startToBreakMin, const unsigned int breakToStopMin, const unsigned int limit, const unsigned int cost) :
    durationMin(durationMin),
    durationMax(durationMax),
    startMin(startMin),
    stopMax(stopMax),
    breakDurationMin(breakDurationMin),
    breakDurationMax(breakDurationMax),
    startToBreakMin(startToBreakMin),
    breakToStopMin(breakToStopMin),
    limit(limit),
    cost(cost),
    vacationList()
{
    if(!check())
    {
        throw invalid_argument("Invalid parameters [VacationType constructor]");
    }
    enumerate();
}

void VacationType::enumerate()
{
    for(unsigned int duration = durationMin; duration <= durationMax; ++duration)
    for(unsigned int start = startMin; start + duration - 1 <= stopMax; ++start)
    for(unsigned int breakDuration = breakDurationMin; breakDuration <= breakDurationMax; ++breakDuration)
    for(unsigned int breakStart = start + startToBreakMin; breakStart + breakDuration + breakToStopMin - 1 <= start + duration - 1; ++breakStart)
    {
        vacationList.push_back(Vacation(start, start + duration - 1, breakStart, breakStart + breakDuration - 1));
    }
}

bool VacationType::check() const
{
    return
           (durationMin <= durationMax)
        && (startMin <= stopMax)
        && (startMin + durationMax - 1 <= stopMax)
        && (breakDurationMin <= breakDurationMax)
        && (startToBreakMin + breakDurationMin + breakToStopMin <= durationMin)
        && (startToBreakMin + breakDurationMax + breakToStopMin <= durationMax);
}

const vector< Vacation > & VacationType::vacations() const
{
    return vacationList;
}

ostream & operator << (ostream & output, const VacationType & vacationType)
{
    output << "(";
    output << "durationMin = " << vacationType.durationMin << ", ";
    output << "durationMax = " << vacationType.durationMax << ", ";
    output << "startMin = " << (vacationType.startMin + 1) << ", ";
    output << "stopMax = " << (vacationType.stopMax + 1) << ", ";
    output << "breakDurationMin = " << vacationType.breakDurationMin << ", ";
    output << "breakDurationMax = " << vacationType.breakDurationMax << ", ";
    output << "startToBreakMin = " << vacationType.startToBreakMin << ", ";
    output << "breakToStopMin = " << vacationType.breakToStopMin << ", ";
    output << "limit = " << vacationType.limit << ", ";
    output << "cost = " << vacationType.cost << ")";
    return output;
}
